### PR TITLE
task/WG-96: support questionnaires with assets

### PR DESCRIPTION
**Pull request by nathanfranklin**
## Overview: ##

Support assets in questionnaires.  

This PR imports the .rq file and all the associated asset images.

## Related Jira tickets: ##

* [WG-96](https://jira.tacc.utexas.edu/browse/WG-96)

## Testing Steps: ##
1. Test with related hazmapper branch https://github.com/TACC-Cloud/hazmapper/pull/144
2. Load questionnaire from the example in https://www.designsafe-ci.org/data/browser/projects/8131157524548227566-242ac117-0001-012/ "PRJ-4019 | Questionnaire_Example"
3. ensure the questionnaire is loaded
4. and can be viewed.

### CODE REVIEW

File: geoapi/services/images.py [hunk 1]
```diff
@@ -79,14 +79,9 @@ def _fix_orientation(fileObj: IO) -> PILImage:
     # from https://github.com/ianare/exif-py#usage-example
```
Comment (nathanfranklin) [inline geoapi/services/images.py:79]: changes here unrelated to PR

File: geoapi/services/features.py [hunk 2]
```diff
@@ -240,15 +239,22 @@ def fromShapefile(database_session, projectId: int, fileObj: IO, metadata: Dict,
...
+        Import RAPP questionnaire
+
+        RAPP questionnaire is imported along with any asset images that it
+        refers to. The asset images are assumed to reside in the same directory
+        as the questionnaire .rq file.
 
         :param projectId: int
-        :param fileObj: file descriptor
-        :param metadata: Dict of <key, val> pairs
+        :param fileObj: questionnaire file
```
Comment (taoteg) [inline geoapi/services/features.py:252]: Where is the `questionnaire_file` value coming from? Its probably just me, but I do not see this defined anywhere before it is used here.

Comment (taoteg) [reply geoapi/services/features.py:252]: Maybe if it specified `questionnaire rq file` it would be clearer?

File: geoapi/services/features.py [hunk 3]
```diff
@@ -264,9 +270,40 @@ def fromRAPP(database_session, projectId: int, fileObj: IO, metadata: Dict, orig
         pathlib.Path(questionnaire_path).mkdir(parents=True, exist_ok=True)
         asset_path = os.path.join(questionnaire_path, 'questionnaire.rq')
 
+        # write questionnaire file (rq)
```
Comment (taoteg) [inline geoapi/services/features.py:273]: This seems to be the only other reference to the `questionnaire file` string and here you specify the `rq` value. I think it is unclear from the base string (`questionnaire file`) that it is rq file and adding that would alleviate the need to specify it here, right?

File: geoapi/tasks/external_data.py [hunk 4]
```diff
@@ -57,47 +66,71 @@ def get_file(client, system_id, path, required):
...
+        logger.info(f'Determining which shapefile-related files need to be downloaded for file {current_file.filename}')
         for extension, required in SHAPEFILE_FILE_ADDITIONAL_FILES.items():
-            additional_file_path = path.with_suffix(extension)
+            additional_file_path = current_file_path.with_suffix(extension)
             if available_files and str(additional_file_path) not in available_files:
                 if required:
-                    logger.error("Could not import required shapefile-related file: "
-                                 "agave: {} :: {}".format(systemId, additional_file_path))
-                    raise Exception("Required file ({}) missing".format(additional_file_path))
+                    logger.error(f"Could not import required shapefile-related file: agave: {system_id}/{additional_file_path}")
```
Comment (taoteg) [inline geoapi/tasks/external_data.py:93]: Should these be single quotes?

File: geoapi/tasks/external_data.py [hunk 5]
```diff
@@ -57,47 +66,71 @@ def get_file(client, system_id, path, required):
...
+    elif file_suffix == "rq":
+        logger.info(f'Parsing rq file {current_file.filename} to see what assets need to be downloaded ')
+        data = json.load(current_file)
+        for section in data["sections"]:
+            for question in section["questions"]:
+                for asset in question.get("assets", []):
+                    # determine full path for this asset and add to list
+                    additional_file_path = current_file_path.with_name(asset["filename"])
+                    additional_files_to_get.append(AdditionalFile(path=additional_file_path, required=True))
+        logger.info(f'{len(additional_files_to_get)} assets were found for rq file {current_file.filename}')
```
Comment (taoteg) [inline geoapi/tasks/external_data.py:107]: There is an inconsistency with using single and double quotes through this file. Could we normalize that as much as possible to one of the two?
Compare lines 88, 93, 94, 99, 107, 122, etc.

Comment (taoteg) [reply geoapi/tasks/external_data.py:107]: I don't know that TACC WMA has a preference but I tend to prefer this usage:

- raw string literals for regular expressions
- single quotes for small symbol-like strings
- double quotes around strings that are used for interpolation or that are natural language messages (break the rules as needed if the strings contain quotes to escape)
- triple double quotes for docstrings

File: geoapi/tests/fixtures/questionnaire.rq
Comment (taoteg) [inline geoapi/tests/fixtures/questionnaire.rq:1]: Should the name be `questionnaire_without_assets.rq` to be explicit and consistent with the other fixture below?

File: geoapi/tests/external_data_tests/test_external_data.py [hunk 6]
```diff
@@ -449,16 +449,16 @@ def test_is_member_of_rapp_project_folder():
...
 
-def test_get_additional_files_none(agave_utils_with_geojson_file):
-    assert not get_additional_files("testSystem", "/testPath/file.jpg", agave_utils_with_geojson_file)
+def test_get_additional_files_none(shapefile_fixture, agave_utils_with_geojson_file):
+    assert not get_additional_files(shapefile_fixture, "testSystem", "/testPath/file.jpg", agave_utils_with_geojson_file)
 
 
-def test_get_additional_files(agave_utils_with_geojson_file):
-    files = get_additional_files("testSystem", "/testPath/file.shp", agave_utils_with_geojson_file)
+def test_get_additional_shapefiles_files(shapefile_fixture, agave_utils_with_geojson_file):
```
Comment (taoteg) [inline geoapi/tests/external_data_tests/test_external_data.py:456]: If 461 and 494 are correct, maybe line 456 needs to be `test_get_additional_files_shapefiles`?

Comment (taoteg) [reply geoapi/tests/external_data_tests/test_external_data.py:456]: Feels like this is the correct change.

File: geoapi/tests/external_data_tests/test_external_data.py [hunk 7]
```diff
@@ -482,17 +483,35 @@ def test_get_additional_files_with_available_files(agave_utils_with_geojson_file
...
+    files = get_additional_files(shapefile_fixture,
+                                 "testSystem",
                                  "/testPath/file.shp",
                                  agave_utils_with_geojson_file,
                                  available_files=available_files)
     assert len(files) == 3
 
 
-def test_get_additional_files_but_missing_prj(agave_utils_with_geojson_file):
+def test_get_additional_files_shapefiles_but_missing_prj(shapefile_fixture, agave_utils_with_geojson_file):
```
Comment (taoteg) [inline geoapi/tests/external_data_tests/test_external_data.py:494]: I would drop the `_but_` in the name.

File: geoapi/tests/external_data_tests/test_external_data.py [hunk 8]
```diff
@@ -482,17 +483,35 @@ def test_get_additional_files_with_available_files(agave_utils_with_geojson_file
...
     with pytest.raises(Exception):
-        get_additional_files("testSystem",
+        get_additional_files(shapefile_fixture,
+                             "testSystem",
                              "/testPath/file.shp",
                              agave_utils_with_geojson_file,
                              available_files=available_files_missing_prj)
+
+
+def test_get_additional_files_rapid_questionnaire(questionnaire_file_with_assets_fixture, agave_utils_with_geojson_file):
```
Comment (taoteg) [inline geoapi/tests/external_data_tests/test_external_data.py:504]: Should this be `test_get_additional_files_rapid_questionnaire_with_assets` to be explicit?

**Review by taoteg — APPROVED** [review]
Approved with a few recommendations on naming choices and style consistency.

File: geoapi/tests/fixtures/questionnaire.rq
Comment (nathanfranklin) [reply geoapi/tests/fixtures/questionnaire.rq:1]: good idea. Fixed 👍

File: geoapi/tests/external_data_tests/test_external_data.py (hunk 8 above)
Comment (nathanfranklin) [reply geoapi/tests/external_data_tests/test_external_data.py:504]: done 👍

File: geoapi/services/features.py (hunk 3 above)
Comment (nathanfranklin) [reply geoapi/services/features.py:273]: yes, i updated here to:
```
        # write questionnaire rq file
```
and above in docstr to:
```
        :param fileObj: questionnaire rq file

```

File: geoapi/tasks/external_data.py (hunk 5 above)
Comment (nathanfranklin) [reply geoapi/tasks/external_data.py]: I took a stab at this in  65eeef84e6a409bbad1d1eda9a241e33f79b77f7.  let me know if that matches up with your comment.

**Review by sophia-massie — APPROVED** [review]
LGTM

**Pull request merged**